include/zephyr/drivers/charger.h

```diff
@@ -0,0 +1,255 @@
+		.type = CHARGER_TYPE_ONLINE, .value = {.online = (_value) }                        \
+	}
+#define CHARGER_VALUE_STATUS_INITIALIZER(_value)                                                   \
+	{                                                                                          \
+		.type = CHARGER_TYPE_STATUS, .value = {.status = (_value) }                        \
+	}
+
+#define CHARGER_ASSERT_TYPE(_v, _type) __ASSERT_NO_MSG((_v)->type == (_type))
+
+/**
+ * @brief A 'get' helper function for boolean values
+ *
+ * @param v charger value container
+ * @retval boolean flag - see CHARGER_TYPE_BOOL
+ */
+static inline bool charger_value_get_bool(const struct charger_value *v)
+{
+	CHARGER_ASSERT_TYPE(v, CHARGER_TYPE_BOOL);
+	return v->value.b;
+}
+
+/**
+ * @brief A 'get' helper function for uint values
+ *
+ * @param v charger value container
+ * @retval uint value - see CHARGER_TYPE_UINT
+ */
+static inline int charger_value_get_uint(const struct charger_value *v)
+{
+	CHARGER_ASSERT_TYPE(v, CHARGER_TYPE_UINT);
+	return v->value.uint;
+}
+
+/**
+ * @brief A 'get' helper function for charger_online values
+ *
+ * @param v charger value container
+ * @return enum charger_online
+ */
+static inline enum charger_online charger_value_get_online(const struct charger_value *v)
+{
+	CHARGER_ASSERT_TYPE(v, CHARGER_TYPE_ONLINE);
+	return v->value.online;
+}
+
+/**
+ * @brief A 'get' helper function for charger_status values
+ *
+ * @param v charger value container
+ * @return enum charger_status
+ */
+static inline enum charger_status charger_value_get_status(const struct charger_value *v)
+{
+	CHARGER_ASSERT_TYPE(v, CHARGER_TYPE_STATUS);
+	return v->value.status;
+}
+
+/**
+ * @typedef charger_get_property_t
+ * @brief Callback API for getting a charger property.
+ *
+ * See charger_get_property() for argument description
+ */
+typedef int (*charger_get_property_t)(const struct device *dev, charger_property_t prop,
+				      struct charger_value *val);
+
+/**
+ * @brief Charging device API
+ *
+ * Caching is entirely on the onus of the client
+ */
+__subsystem struct charger_driver_api {
+	charger_get_property_t get_property;
+};
+
+/**
+ * @brief Fetch a battery charger property
+ *
+ * @param dev Pointer to the battery charger device
+ * @param prop Charger property identifier - see charger_property_t
+ * @param val Pointer to a struct for loading the gotten prop value
+ *
+ * @retval 0 if successful
+ * @retval < 0 if getting property failed
+ */
+__syscall int charger_get_prop(const struct device *dev, charger_property_t prop,
+			       struct charger_value *val);
+
+static inline int z_impl_charger_get_prop(const struct device *dev, charger_property_t prop,
+					  struct charger_value *val)
+{
+	const struct charger_driver_api *api = (const struct charger_driver_api *)dev->api;
+	int ret;
+
+	ret = api->get_property(dev, prop, val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	switch (prop) {
```

Comment:
Why is this switch needed?

Reply:
how else would you like to see the `.type` assigned?

I thought this was necessary to perform the type assertion. Also FYI this is from the API level moving to the driver level within the hour.